cuqi/experimental/mcmc/_hmc.py

```diff
@@ -212,9 +216,9 @@ def step(self):
         self._num_tree_node = 0
 
         # copy current point, logd, and grad in local variables
-        point_k = self.current_point.copy() # initial position (parameters)
+        point_k = deepcopy(self.current_point) # initial position (parameters)
```

Comment:
Are you sure we need this deep copy? One may try a more efficient operation like the "slicing trick" self.current_point[:]. We can wrap this in a method for NUTS like "self._copy_array()" or something.

it would be good to have a bit of explanation as to why it is needed.

Reply:
Thank you @nabriis , very good point. I think deep copy made a bit of an overhead, maybe not significant and also not significant relative to computation time. For example the Gibbs example used to run in 19.6 Sec but after the modification with deep copy it runs in 20.7 Sec.

To address this, I revert to using numpy.copy which I read is more efficient than deep copy. I handelded cases where the amount could be scalar or array with if else.

The reason I used deepcopy at the first place is because some values like logd sometimes is array and sometimes is scalar.